### ISSUE TITLE
Configurable `Query` class to be used for operation evaluation

### ISSUE THREAD

**Issue by ravangen** [rmosolgo/graphql-ruby#4354]
**Is your feature request related to a problem? Please describe.**

Similar to [`context_class`](https://github.com/rmosolgo/graphql-ruby/blob/15fec5f50da50f990571be51f0e8e6e013cbaee6/lib/graphql/schema.rb#L731-L737) on `Schema`, allow for a custom `Query` class to be defined and used as part of the execution process.

**Describe alternatives you've considered**

I have a workaround where I override `multiplex` to use a custom subclass of `Query`.

```ruby
def multiplex(queries, **kwargs)
  queries = queries.map do |query|
    case query
    when Hash
      query_class.new(self, nil, **query)
    else
      query
    end
  end
  super(queries, **kwargs)
end
```

**Comment by rmosolgo** [comment]
I'm certainly open to it, but, may I ask why?

**Comment by ravangen** [comment]
Sure, we are going through some refactoring tweaks, so it isn't critical, but made us think about `context_class`.

Today we have a class that wraps (delegates) to `GraphQL::Query` and calls `result` on it. As part of the wrapping, we have a bunch of business logic. I was thinking that perhaps one or two of the following concerns could live within a subclass of `GraphQL::Query`:
1. exception handling/classification
2. response caching (e.g. json)
3. operation caching (remember static validation outcome of a past run for same query document)
4. black hole specific query documents
5. some legacy things

We have a custom internal gem that wraps `graphql` that seeks to provide some common abstractions for teams across codebases to use based on some internal conventions and common tooling we have. So we will have some form of encapsulation.

I was making some observability instrumentation tweaks, and I had some shared code (metric tags/labels) between instrumentation and tracer classes that I wanted to lazily evaluate once per request. Normally I use `context` to hold this state, but personally I feel it can be a overused dumping ground of state. What triggered this question was me playing around with putting this state on the subclass vs storing a `proc` in `context`, that holds an instance variable of a hash.

```ruby
def metric_tags
  @metric_tags ||= {
    schema: schema.handle,
    operation_type: selected_operation.operation_type,
    operation_name: metric_operation_name,
    api_client: metric_api_client,
  }
end
```

```ruby
context = {
  # ...
  metric_tags: Proc.new do
    @metric_tags ||= {
      schema: schema.handle,
      operation_type: selected_operation.operation_type,
      operation_name: metric_operation_name, # handle cardinality concerns with an allowlist
      api_client: metric_api_client, # handle cardinality concerns with an allowlist
    }
  end,
}

context[:metric_tags].call
```

For the information I want to put in `metrics_tags`, a lot of it is query state driven.

**Comment by rmosolgo** [comment]
> dumping ground


To be fair, wouldn't a configurable Query class just make _both_ `context` and `query` into dumping grounds for state? 

Or, to say it another way, what about implementing `def metric_tags` inside a custom context class? 

```
def metric_tags
  @metric_tags ||= {
    schema: query.schema.handle, 
    # ...
  }
end 
```

You can always get the query via `context.query`, so I think they should be practically equivalent. What do you think of that? 

In any case, it doesn't really add much complexity to the library to support this. I tried it out here, and if it looks good to you, we can go for it: https://github.com/rmosolgo/graphql-ruby/pull/4679

**Comment by ravangen** [comment]
Agreed, the `metrics_tags` is a poor example because it could be in a custom `Context`, and accessed via the `Query` instance (e.g. `context.metric_tags`, `query.context.metric_tags`, `result.query.context.metric_tags`). It does have a feeling of state and contains values like operation name and api client id.

We do override the default `Query` behaviour for things closer to its concerns to achieve things like:
- exception handling/classification
- operation caching (remember static validation outcome of a past run for same query document)
- black hole specific query documents
